apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+
+        address submitter = action.submitter;
+        require(msg.sender == submitter, ERROR_SENDER_NOT_ALLOWED);
+
+        (,, IArbitrator arbitrator) = _getSettingFor(action);
+        bytes memory metadata = _buildDisputeMetadata(_actionId);
+        uint256 disputeId = _createDispute(action, challenge, arbitrator, metadata);
+        _submitEvidence(arbitrator, disputeId, submitter, action.context, _submitterFinishedEvidence);
+        _submitEvidence(arbitrator, disputeId, challenge.challenger, challenge.context, challenge.challengerFinishedEvidence);
+
+        challenge.state = ChallengeState.Disputed;
+        challenge.disputeId = disputeId;
+        challenge.submitterFinishedEvidence = _submitterFinishedEvidence;
+        challengeByDispute[disputeId] = challengeId;
+        emit ActionDisputed(_actionId, challengeId);
+    }
+
+    /**
+    * @notice Submit evidence for dispute #`_disputeId`
+    * @param _disputeId Identification number of the dispute on the arbitrator
+    * @param _evidence Evidence data submitted for the dispute
+    * @param _finished Whether the submitter is finished submitting evidence
+    */
+    function submitEvidence(uint256 _disputeId, bytes _evidence, bool _finished) external {
+        (uint256 _actionId, Action storage action, , Challenge storage challenge) = _getDisputedAction(_disputeId);
+        require(_isDisputed(_actionId, challenge), ERROR_CANNOT_SUBMIT_EVIDENCE);
+
+        (, , IArbitrator arbitrator) = _getSettingFor(action);
+        bool submitterAndChallengerFinished = _updateEvidenceSubmissionStatus(action, challenge, msg.sender, _finished);
+        _submitEvidence(arbitrator, _disputeId, msg.sender, _evidence, _finished);
+
+        if (submitterAndChallengerFinished) {
+            arbitrator.closeEvidencePeriod(_disputeId);
+        }
+    }
+
+    /**
+    * @notice Rule the action associated to dispute #`_disputeId` with ruling `_ruling`
+    * @param _disputeId Identification number of the dispute on the arbitrator
+    * @param _ruling Ruling given by the arbitrator
+    */
+    function rule(uint256 _disputeId, uint256 _ruling) external {
+        (uint256 actionId, Action storage action, uint256 challengeId, Challenge storage challenge) = _getDisputedAction(_disputeId);
+        require(_canRuleDispute(actionId, challenge), ERROR_CANNOT_RULE_ACTION);
+
+        (, , IArbitrator arbitrator) = _getSettingFor(action);
+        require(arbitrator == IArbitrator(msg.sender), ERROR_SENDER_NOT_ALLOWED);
+
+        challenge.ruling = _ruling;
+        emit Ruled(arbitrator, _disputeId, _ruling);
+
+        // TODO: implement try catch
+        if (_ruling == DISPUTES_RULING_SUBMITTER) {
+            _rejectChallenge(actionId, action, challengeId, challenge);
+        } else if (_ruling == DISPUTES_RULING_CHALLENGER) {
+            _acceptChallenge(actionId, action, challengeId, challenge);
+        } else {
+            _voidChallenge(actionId, action, challengeId, challenge);
```

Comment:
Would it be better to revert on all other `ruling`s that don't match our expected results?

Or is it better to just leave them all as `void`?

Reply:
I think it's safe to understand it as an invalid ruling (refuse - void)

Reply:
I'll add it to the list of assumptions